src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,232 @@
+{
+	repl_origin_local_session_start_no_publish();
+	PG_RETURN_VOID();
+}
+
+
+Datum
+replication_origin_session_end_no_publish(PG_FUNCTION_ARGS)
+{
+	repl_origin_local_session_end_no_publish();
+	PG_RETURN_VOID();
+}
+
+
+Datum
+replication_origin_session_is_no_publish(PG_FUNCTION_ARGS)
+{
+	bool result = repl_origin_local_session_is_no_publish();
+	PG_RETURN_BOOL(result);
+}
+
+
+void
+repl_origin_local_session_start_no_publish(void)
+{
+	originalOriginId = replorigin_session_origin;
+	replorigin_session_origin = DoNotReplicateId;
+}
+
+
+void
+repl_origin_local_session_end_no_publish(void)
+{
+	replorigin_session_origin = originalOriginId;
+}
+
+
+bool
+repl_origin_local_session_is_no_publish(void)
+{
+	return (replorigin_session_origin != InvalidRepOriginId);
+}
+
+
+/*
+ * SetupReplicationOriginRemoteSession sets up a new replication origin session in a
+ * remote session. The identifier is used to create a unique replication origin name
+ * for the session in the remote node.
+ */
+void
+SetupReplicationOriginRemoteSession(MultiConnection *connection)
+{
+	/*elog(LOG, "Setting up remote replication origin session %p %s,%d", connection, connection->hostname, connection->port); */
+	if (connection != NULL && !IsReplicationOriginSessionSetup(connection))
+	{
+		/*elog(LOG, "After IsReplicationOriginSessionSetup session %s,%d", connection->hostname, connection->port); */
+		StringInfo replicationOriginSessionSetupQuery = makeStringInfo();
+		appendStringInfo(replicationOriginSessionSetupQuery,
+						 "select pg_catalog.replication_origin_session_start_no_publish();");
+		ExecuteCriticalRemoteCommand(connection,
+									 replicationOriginSessionSetupQuery->data);
+		connection->isReplicationOriginSessionSetup = true;
+
+		/*elog(LOG, "After Setting up remote replication origin session %s,%d", connection->hostname, connection->port); */
+	}
+}
+
+
+/*
+ * ResetReplicationOriginRemoteSession resets the replication origin session in a
+ * remote node.
+ */
+void
+ResetReplicationOriginRemoteSession(MultiConnection *connection)
+{
+	if (connection != NULL && connection->isReplicationOriginSessionSetup)
+	{
+		/*elog(LOG, "Resetting remote replication origin session %s,%d", connection->hostname, connection->port); */
+
+		/* Reset the replication origin in remote session*/
+		StringInfo replicationOriginSessionResetQuery = makeStringInfo();
+		appendStringInfo(replicationOriginSessionResetQuery,
+						 "select pg_catalog.replication_origin_session_end_no_publish();");
+		ExecuteCriticalRemoteCommand(connection,
+									 replicationOriginSessionResetQuery->data);
+		connection->isReplicationOriginSessionSetup = false;
+	}
+}
+
+
+/*
+ * SetupReplicationOriginLocalSession sets up a new replication origin session in a
+ * local session.
+ */
+void
+SetupReplicationOriginLocalSession(void)
+{
+	/*elog(LOG, "Setting up local replication origin session"); */
+	if (!IsReplicationOriginSessionSetup(NULL))
```

Comment:
can we add a setting here to put all replication origins assignments behind a feature flag?

Reply:
OK. Added a feature flag "citus.enable_replication_origin_session" for the UDFs and all replication origin functions that are called from other files in code.